### PR TITLE
Johnc ibdprune

### PR DESCRIPTION
**Pull request by johnc1231**
Added IBD prune functionality to VDS's.

### CODE REVIEW

File: python/hail/dataset.py [hunk 1]
```diff
@@ -2014,6 +2014,34 @@ def ibd(self, maf=None, bounded=True, min=None, max=None):
...
+        >>> vds.ibd_prune(.4)
+
+        **Notes**
+
+        As this is still actively being developed, dataset returned by this function may change in near future
+        as a result of algorithmic improvements. As it currently stands, this algorithm will work very well on
+        tons of small families, but be rather slow on fewer but larger families.
+
+
+        :param threshold: The minimum PI_HAT value between samples that will be present in the pruned dataset.
```
Comment (johnc1231) [inline python/hail/dataset.py]: I think it's clearer, going with it.

File: src/main/scala/is/hail/methods/MaximalIndependentSet.scala [hunk 2]
```diff
@@ -40,42 +35,45 @@ object MaximalIndependentSet {
...
-    val numberedVertices: RDD[(String, VertexId)] = allVertices.zipWithIndex()
+      filteredGraph.unpersist()
+    }
+    originalGraph.subgraph(_ => true, (id, value) => !idSetToDelete.contains(id)).vertices.keys.collect().toSet
+  }
 
-    val verticesToNumbers = numberedVertices.collectAsMap()
-    val numbersToVertices = numberedVertices.map{case (name, id) => (id, name)}.collectAsMap()
 
+  def ofIBDMatrix(inputRDD: RDD[((Int, Int), Double)], thresh: Double, vertexIDs: Seq[Int]): Set[Long] = {
```
Comment (johnc1231) [inline src/main/scala/is/hail/methods/MaximalIndependentSet.scala]: We cannot. In the line 

`val edges: RDD[Edge[Double]] = filteredRDD.map{case((v1, v2), weight) => Edge(v1, v2, weight)}`

`v1` and `v2` are the integers being used as ids for the vertices. Technically longs would also work, but only ints or longs.

File: src/main/scala/is/hail/methods/MaximalIndependentSet.scala [hunk 3]
```diff
@@ -40,42 +35,45 @@ object MaximalIndependentSet {
...
-    //Collect all vertices.
-    val allVertices = vertexPairs.flatMap[String]{case (v1, v2) => List(v1, v2)}.distinct()
+      idSetToDelete = idSetToDelete.union(pregelGraph.vertices
+        .filter(tuple => tuple match {case (id, value) => value match  {case (maxDegrees, maxID) => maxID == id && maxDegrees != 0}})
+        .map(_._1).collect().toSet)
 
-    val numberedVertices: RDD[(String, VertexId)] = allVertices.zipWithIndex()
+      filteredGraph.unpersist()
+    }
+    originalGraph.subgraph(_ => true, (id, value) => !idSetToDelete.contains(id)).vertices.keys.collect().toSet
```
Comment (johnc1231) [inline src/main/scala/is/hail/methods/MaximalIndependentSet.scala]: Agreed, this is simpler. Making changes accordingly.

**Comment by johnc1231** [comment]
I think that's everything fixed, back to you (after you get back from vacation of course)

**Comment by jbloom22** [comment]
@johnc1231, re-open this when Dan is back from vacation

File: src/main/scala/is/hail/methods/MaximalIndependentSet.scala [hunk 4]
```diff
@@ -40,42 +36,45 @@ object MaximalIndependentSet {
...
+          .pregel(initialMsg, Int.MaxValue, edgeDirection)(receiveMessage, sendMsg, mergeMsg)
+          .vertices
+          .filter{ case (id, (maxDegrees, maxID)) => maxID == id && maxDegrees != 0}
+          .map(_._1)
+          .collect()
+          .toSet
+      workingGraph.unpersist()
+      workingGraph = g.subgraph(_ => true, (id, value) => !idSetToDelete.contains(id))
+      workingGraph.persist()
+    } while (workingGraph.numEdges > 0)
```
Comment (danking) [inline src/main/scala/is/hail/methods/MaximalIndependentSet.scala]: I think I lead you astray, I'm sorry! We actually want the condition at the top of the loop, i.e. a standard while loop. While this is correct, I think we end up with one unnecessary iteration if the graph is completely disconnected, right?

**Comment by danking** [comment]
@johnc1231 Could you also time this routine on the european subset of 1000KG? with PI_HAT=0.2? I'd like to have that information in the PR comments in case we need to know that in the future.

**Comment by johnc1231** [comment]
I fixed that last thing, I will add benchmark comment when I have it.

**Pull request merged**